File: Session14/Java/DijkstraAlgorithm.java
https://practice.geeksforgeeks.org/problems/implementing-dijkstra-set-1-adjacency-matrix/1

class Solution
{
    static class Pair {
        int weight;
        int node;

        public Pair(int weight, int node) {
            this.weight = weight;
            this.node = node;
        }
    }
    //Function to find the shortest distance of all the vertices
    //from the source vertex S.
    static int[] dijkstra(int V, ArrayList<ArrayList<ArrayList<Integer>>> adj, int S)
    {
        // Write your code here
        PriorityQueue<Pair> pq = new PriorityQueue<>(Comparator.comparingInt(p -> p.weight));
        // PriorityQueue<Pair> pq = new PriorityQueue<>((a,b) -> (a.weight - b.weight));
        pq.add(new Pair(0, S));
        int[] distance = new int[V];
        Arrays.fill(distance, Integer.MAX_VALUE);
        distance[S] = 0;

        while (!pq.isEmpty()) {
            int node = pq.peek().node;
            int weight = pq.peek().weight;
            pq.poll();

            for (ArrayList<Integer> nbr : adj.get(node)) {
                int adjNode = nbr.get(0);
                int adjWeight = nbr.get(1);
                if (distance[adjNode] > weight + adjWeight) {
                    distance[adjNode] = weight + adjWeight;
                    pq.add(new Pair(distance[adjNode], adjNode));
                }
            }
        }
        return distance;
    }
}

File: Session14/C++/DijkstraAlgorithm.cpp
https://practice.geeksforgeeks.org/problems/implementing-dijkstra-set-1-adjacency-matrix/1

    vector <int> dijkstra(int V, vector<vector<int>> adj[], int S)
    {
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq; // min heap
        // distance, node
        pq.push({0, S});
        vector<int> distance(V, INT_MAX);
        distance[S] = 0;
    
        while (!pq.empty()) {
            int node = pq.top().second;
            int weight = pq.top().first;
            pq.pop();
            
            for (auto nbr : adj[node]) {
                int adjNode = nbr[0];
                int adjWeight = nbr[1];
                if (distance[adjNode] > weight + adjWeight) {
                    distance[adjNode] = weight + adjWeight;
                    pq.push({distance[adjNode], adjNode});
                }
            }
        }
        return distance;
    }